doc/lua_api.md

```diff
@@ -5466,21 +5466,44 @@ Call these functions only at load time!
     * Returns an integer object handle uniquely identifying the registered
       ore on success.
     * The order of ore registrations determines the order of ore generation.
+* `minetest.override_ore(name, ore definition)`
+    * Override existing ore.
+* `minetest.unregister_ore(name)`
+    * Unregisters the decoration from the engine, and deletes the entry with key
+      `name` from `minetest.registered_decorations`.
+* `minetest.reregister_ores()`
+    * Clears all ores currently registered and register them again.
+    * Can be used after overriding/unregistering biome/bimes.
```

Comment:
Such re-registering functions should not need to exist. Mods would end up calling it all the time after modifying a registration. Couldn't you do a direct call to C++ side with each `minetest.override_ore`?

Reply:
This function should be used after some changes in registered biomes.
There is no need to call it after changes in registered ores.

I can remove it, and call it after override_biome and unregister_biome calls. Does it sound ok?

Reply:
Do biomes change their internal ID ( `biome->index`) when they're modified? I cannot see any other internal relation aside from that. For example, node names are resolved after `on_mods_loaded`, so that cannot be the issue.

Reply:
Changing or unregistering the biome is possible only by clearing all biomes and registering them now.
So there is no guarantee here, that biome ID will be the same after that action.

It is documented:
https://github.com/minetest/minetest/blob/6952bab51993df33c59b0987799d19590f9ba854/doc/lua_api.md?plain=1#L5528-L5532

It is the reason for reregistering functions to be implemented.

Reply:
So it appears that the ore registration should not resolve the biome name --> ID mapping immediately but together this the node resolver. That way such workarounds would no longer be necessary.

Reply:
At the actual master branch registration of ore and decoration process everything, also node resolve.

https://github.com/minetest/minetest/blob/6952bab51993df33c59b0987799d19590f9ba854/src/script/lua_api/l_mapgen.cpp#L1237

https://github.com/minetest/minetest/blob/6952bab51993df33c59b0987799d19590f9ba854/src/script/lua_api/l_mapgen.cpp#L1456

Both `core.registration_ore` and `core.registration_decoration` also returned the handle of registered ore/decoration.

If I register biomes, ores, and decorations to C++ mapgen after `on_mods_loaded` call, it will not be possible for `core.registration_ore` and `core.registration_decoration` to return a valid handler. 

It will break backward compatibility probably.

Is it acceptable?

Reply:
The only problem that I can see with `master` is the following code:
https://github.com/minetest/minetest/blob/492aab20feb463888aa19d576d5b239868ab312e/src/script/lua_api/l_mapgen.cpp#L456-L472

It should yet not map ores to biome IDs. Moving it somewhere to the NodeResolver step would allow you to:
1. continue using handles
2. modify, delete and create biomes without worrying about ores

Reply:
I see it now. Thank you for your help. I will try to simplest BiomeResolver class.

Reply:
I just noticed that biomes may also have an empty name, which somewhat breaks my plans. In that case we'd need another reliable identifier...

Reply:
The problem can be also this: 
https://github.com/minetest/minetest/blob/57de599a292d1c7ad43dfc260d36dff2fbb61481/doc/lua_api.md?plain=1#L9908-L9912

Looks like it is allowed in the definition of ore to use biome IDs also. Similar to the definition of decoration.

We should mention this in doc, it should be the modder's responsibility to fix biome IDs in ore/decorations after removing/changing the biome.